packages/webapp/src/components/Task/AddSoilAmendmentProducts/QuantityApplicationRate/useQuantityApplicationRate.ts

```diff
@@ -0,0 +1,151 @@
+ *  (at your option) any later version.
+ *
+ *  LiteFarm is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+
+import { useEffect, useState } from 'react';
+import { TASK_PRODUCT_FIELD_NAMES, UnitOption } from '../types';
+import { getUnitOptionMap } from '../../../../util/convert-units/getUnitOptionMap';
+import { convertFn, getDefaultUnit, location_area } from '../../../../util/convert-units/unit';
+import { roundToTwoDecimal } from '../../../../util';
+import { convert } from '../../../../util/convert-units/convert';
+import { useFormContext } from 'react-hook-form';
+
+interface UseQuantityApplicationRate {
+  total_area: number;
+  total_area_unit?: 'm2' | 'ha' | 'ft2' | 'ac'; // as defined in location_area
+  system: 'metric' | 'imperial';
+  isWeight: boolean;
+  namePrefix: string;
+}
+
+export const useQuantityApplicationRate = ({
+  total_area,
+  total_area_unit,
+  system,
+  isWeight,
+  namePrefix,
+}: UseQuantityApplicationRate) => {
+  const TOTAL_AREA_AMENDED = `${namePrefix}.${TASK_PRODUCT_FIELD_NAMES.TOTAL_AREA_AMENDED}`;
+  const TOTAL_AREA_AMENDED_UNIT = `${namePrefix}.${TASK_PRODUCT_FIELD_NAMES.TOTAL_AREA_AMENDED_UNIT}`;
+  const VOLUME = `${namePrefix}.${TASK_PRODUCT_FIELD_NAMES.VOLUME}`;
+  const WEIGHT = `${namePrefix}.${TASK_PRODUCT_FIELD_NAMES.WEIGHT}`;
+  const APPLICATION_RATE_WEIGHT = `${namePrefix}.${TASK_PRODUCT_FIELD_NAMES.APPLICATION_RATE_WEIGHT}`;
+  const APPLICATION_RATE_VOLUME = `${namePrefix}.${TASK_PRODUCT_FIELD_NAMES.APPLICATION_RATE_VOLUME}`;
+
+  const { setValue, getValues, watch } = useFormContext();
+
+  const application_area_unit = watch(TOTAL_AREA_AMENDED_UNIT);
+
+  /* Update application area + rate based on percent of location */
+  const onPercentLocationChange = (percent_of_location: number) => {
+    const calculatedArea = (total_area * Math.min(percent_of_location || 100, 100)) / 100;
+    setValue(TOTAL_AREA_AMENDED, calculatedArea);
+
+    /* set unit of the total area component according to default breakpoints */
+    setValue(
+      TOTAL_AREA_AMENDED_UNIT,
+      /* @ts-ignore */
+      getUnitOptionMap()[getDefaultUnit(location_area, calculatedArea, system).displayUnit],
+      { shouldValidate: true },
+    );
+
+    updateApplicationRate();
+  };
+
+  const updateApplicationRate = (updatedUnit?: UnitOption) => {
+    const weight = getValues(WEIGHT);
+    const volume = getValues(VOLUME);
+    const application_area = getValues(TOTAL_AREA_AMENDED);
+
+    if ((weight || weight === 0) && isWeight && application_area) {
+      setValue(
+        APPLICATION_RATE_WEIGHT,
+        convert(weight / application_area)
+          .from(isWeight && updatedUnit ? updatedUnit.value : 'kg/m2') // database or selected unit weight / database unit area
```

Comment:
Sorry for nitpicking 😁 but this can be simplified to `updatedUnit?.value || 'kg/m2'.` The `isWeight `check is already done at the top in the if condition. Same for else if below.

Reply:
Oh thank you, I will change it!

Reply:
Hmm, unfortunately I don't think it's going to work. As we are on quite a time crunch I'm going to just revert the handler change, but the changes needed to the calculations to work like this are actually quite a lot.

For instance, the `updateUnit` being passed by the `onChangeUnitOption` handler to `updateApplicationRate` is going to be in e.g. `lbs`, let's say for an imperial farm. But the value being pulled by `getValue()` is in 'kg/m2'.

I don't think there is clean logic for relating a change of `lb` --> `oz` on a metric-based hidden input value without re-writing Unit, which is... doesn't look worth it for avoiding two useEffects. So I'll return to the original method 😓

Reply:
I think that's just the side effect of having too much logic inside of useEffects in `Unit`. Actually re-writing Unit might not be a bad idea because if we're going to be using NumberInput going forward (with the localized numbers), then it probably makes sense to have the numbers in `Unit` behave the same way. Otherwise you have a situation with Unit and NumberInput on the same page behaving differently (commas and periods).